services/venice-controller/src/main/java/com/linkedin/venice/controller/UnusedValueSchemaCleanupService.java

```diff
@@ -0,0 +1,127 @@
+  private final ScheduledExecutorService executor =
+      Executors.newSingleThreadScheduledExecutor(new DaemonThreadFactory("UnusedValueSchemaCleanupService"));
+  private final VeniceControllerMultiClusterConfig multiClusterConfig;
+  private final VeniceParentHelixAdmin veniceParentHelixAdmin;
+  private final VeniceHelixAdmin admin;
+  private final int scheduleIntervalMinutes;
+  private boolean stop = false;
+  public final int minSchemaCountToKeep;
+
+  UnusedValueSchemaCleanupService(
+      VeniceControllerMultiClusterConfig multiClusterConfig,
+      VeniceHelixAdmin admin,
+      VeniceParentHelixAdmin parentHelixAdmin) {
+    this.multiClusterConfig = multiClusterConfig;
+    this.admin = admin;
+    this.scheduleIntervalMinutes = multiClusterConfig.getUnusedSchemaCleanupIntervalMinutes();
+    this.minSchemaCountToKeep = multiClusterConfig.getMinSchemaCountToKeep();
+    this.veniceParentHelixAdmin = parentHelixAdmin;
+  }
+
+  private Runnable getRunnableForSchemaCleanup() {
+    return () -> {
+      if (stop) {
+        return;
+      }
+      for (String clusterName: multiClusterConfig.getClusters()) {
+        boolean cleanupEnabled =
+            multiClusterConfig.getControllerConfig(clusterName).isUnusedValueSchemaCleanupServiceEnabled();
+        if (!cleanupEnabled) {
+          continue;
+        }
+        // Get all stores for current cluster
+        List<Store> stores = admin.getAllStores(clusterName);
+        for (Store store: stores) {
+          String storeName = store.getName();
+          // Remove schema only for batch stores
+          if (isHybridStore(store) || VeniceSystemStoreUtils.isSystemStore(storeName)) {
+            continue;
+          }
+          Collection<SchemaEntry> allSchemas = veniceParentHelixAdmin.getValueSchemas(clusterName, storeName);
+
+          if (allSchemas.size() < minSchemaCountToKeep) {
+            continue;
+          }
+
+          Set<Integer> usedSchemaSet = veniceParentHelixAdmin.getInUseValueSchemaIds(clusterName, storeName);
+
+          if (usedSchemaSet.isEmpty()) {
+            continue;
+          }
+          int minSchemaIdInUse = Collections.min(usedSchemaSet);
+          Set<Integer> schemasToDelete = new HashSet<>();
+
+          // assumes `getValueSchemas` returns ascending schema ids so that the older schemas are deleted first
+          for (SchemaEntry schemaEntry: allSchemas) {
+            int schemaId = schemaEntry.getId();
+            // skip latest value schema or super-set schema id
+            if (schemaId == store.getLatestSuperSetValueSchemaId() || admin.getHelixVeniceClusterResources(clusterName)
+                .getSchemaRepository()
+                .getSupersetOrLatestValueSchema(storeName)
+                .getId() == schemaId) {
+              continue;
+            }
+
+            // delete only if its not used and less than minimum of used schema id
+            if (!usedSchemaSet.contains(schemaId) && schemaId < minSchemaIdInUse) {
```

Comment:
IIUC this is to prevent deletion of new schemas that the user might have added but haven't got a chance to push data with it right? Can we have an unit test for this specific scenario too?

Reply:
I still don't see the unit test include a scenario for this, did I miss it?